Super Table functions done using Spark operations

Retr0Metal98 opened the pull request:
## Changes in this PR

- [ ] Bug fix
- [x] New feature
- [ ] Breaking change of an existing feature
- [ ] Refactor/Performance enhancements

## Overview

- Converted get_files_dynamic.py from model team to do most dataframe ops using Spark DataFrames.
- for now supertable is written to local directory sT.csv, to look into pushing to redshift etc.
- Spark config is done at the start of super_table_pyspark.py and should be adjusted according to cluster configuration.

Retr0Metal98 committed:
Super Table functions done using Spark operations

Converted get_files_dynamic.py from model team to do most dataframe ops using Spark DataFrames.

for now supertable is written to local directory sT.csv, to look into pushing to redshift etc.

Spark config is done at the start of super_table_pyspark.py and should be adjusted according to cluster configuration.

don-tay commented on super_table_pyspark.py:
```diff
@@ -0,0 +1,547 @@
+def get_super_table(call_timestamp=None):
+
+    # Camera to locations -
+    # TODO: perhaps run mapping function one time & pull from S3 instead
```
Is this mapping attempting to de-normalize the row result for camera?

Retr0Metal98 replied:
the mapping function tags each camera with the nearest weather station, this function was already called previously and this mapping is saved in the bucket as 'camera_station_mapping.csv', the updated code just pulls this csv and uses it directly.

The schema mismatch errors have been resolved in the latest commits.

don-tay commented on super_table_pyspark.py:
```diff
@@ -0,0 +1,547 @@
+
+from math import sin, cos, sqrt, atan2, radians
+
+from pyspark.sql import SparkSession
+from pyspark.sql.functions import *
+from pyspark.sql.types import *
+from pyspark.sql.window import Window
```
Do add the dependencies installed from pip into requirements.txt
In the dir, run 
```
pip freeze > requirements.txt
```

don-tay reviewed (changes requested):
I tried running super_table_pyspark.py and got the following error:

```
Traceback (most recent call last):
  File "/Users/dontay/Desktop/traffic-predict/super_table_pyspark.py", line 547, in <module>
    sT = get_super_table()
  File "/Users/dontay/Desktop/traffic-predict/super_table_pyspark.py", line 488, in get_super_table
    non_rainfall_realtime = proc_non_rainfall(call_timestamp=call_timestamp)
  File "/Users/dontay/Desktop/traffic-predict/super_table_pyspark.py", line 292, in proc_non_rainfall
    non_rainfall_realtime = get_df_from_csv(file_path, weather_dir, schema=schema)
  File "/Users/dontay/Desktop/traffic-predict/super_table_pyspark.py", line 100, in get_df_from_csv
    df = spark.createDataFrame(df, schema=schema)
  File "/Users/dontay/Desktop/traffic-predict/.venv/lib/python3.10/site-packages/pyspark/python/lib/pyspark.zip/pyspark/sql/session.py", line 673, in createDataFrame
  File "/Users/dontay/Desktop/traffic-predict/.venv/lib/python3.10/site-packages/pyspark/python/lib/pyspark.zip/pyspark/sql/pandas/conversion.py", line 340, in createDataFrame
  File "/Users/dontay/Desktop/traffic-predict/.venv/lib/python3.10/site-packages/pyspark/python/lib/pyspark.zip/pyspark/sql/session.py", line 700, in _create_dataframe
  File "/Users/dontay/Desktop/traffic-predict/.venv/lib/python3.10/site-packages/pyspark/python/lib/pyspark.zip/pyspark/sql/session.py", line 509, in _createFromLocal
  File "/Users/dontay/Desktop/traffic-predict/.venv/lib/python3.10/site-packages/pyspark/python/lib/pyspark.zip/pyspark/sql/session.py", line 682, in prepare
  File "/Users/dontay/Desktop/traffic-predict/.venv/lib/python3.10/site-packages/pyspark/python/lib/pyspark.zip/pyspark/sql/types.py", line 1411, in verify
  File "/Users/dontay/Desktop/traffic-predict/.venv/lib/python3.10/site-packages/pyspark/python/lib/pyspark.zip/pyspark/sql/types.py", line 1392, in verify_struct
  File "/Users/dontay/Desktop/traffic-predict/.venv/lib/python3.10/site-packages/pyspark/python/lib/pyspark.zip/pyspark/sql/types.py", line 1411, in verify
  File "/Users/dontay/Desktop/traffic-predict/.venv/lib/python3.10/site-packages/pyspark/python/lib/pyspark.zip/pyspark/sql/types.py", line 1405, in verify_default
  File "/Users/dontay/Desktop/traffic-predict/.venv/lib/python3.10/site-packages/pyspark/python/lib/pyspark.zip/pyspark/sql/types.py", line 1293, in verify_acceptable_types
TypeError: field wind_dir_realtime: DoubleType can not accept object 251 in type <class 'int'>
```

don-tay committed:
Add spark dependencies to requirements.txt

don-tay committed:
Add sT folder to gitignore

don-tay committed:
Fix invalid spark df schema to be inferred instead

Retr0Metal98 committed:
Merge pull request #16 from don-tay/fix/invalid-schema

Fix invalid spark df schema to be inferred instead

Retr0Metal98 committed:
Update super_table_pyspark.py

Retr0Metal98 committed:
Update super_table_pyspark.py

fix bug with bing data ingestion:
 - trafficCongestion field is incorrectly inferred by spark as DoubleType.
- correct schema used instead of inferring

don-tay reviewed (approved):
LGTM
